lib/active_model/serializer.rb

```diff
@@ -112,10 +112,24 @@ def self.digest_caller_file(caller_line)
       Digest::MD5.hexdigest(serializer_file_contents)
     end
 
+    def self.serializer_lookup_chain_for(klass)
+      chain = []
+
+      resource_class_name = klass.name.demodulize
+      resource_namespace = klass.name.deconstantize
+      serializer_class_name = "#{resource_class_name}Serializer"
+
+      chain.push("#{name}::#{serializer_class_name}") if self != ActiveModel::Serializer
+      chain.push("#{resource_namespace}::#{serializer_class_name}")
+      chain.push(serializer_class_name)
+
+      chain
+    end
+
     def self.get_serializer_for(klass)
       serializers_cache.fetch_or_store(klass) do
-        serializer_class_name = "#{klass.name}Serializer"
-        serializer_class = serializer_class_name.safe_constantize
+        # NOTE(beauby): When we drop 1.9.3 support we can lazify the map for perfs.
+        serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x }
```

Comment:
> you don't like the name of the method that computes the lookup chain

or I don't like the name of the method that build a list of possible serializer class names :)

> 2. you dislike the use of functional operators and

where?  we don't have lazy and the performance of the each with break should be about the same.  I have to admin that I don't understand a lot of your intentions when it comes to functional programming.  I'd love to be schooled

> 3. you'd rather have this in a separate object.

maybe eventually probably.  that's why I wanted to be conservative and keep it all together for now.

> you seem to dislike the concept of "lookup chain", any specific reason?

it doesn't tell me what it's doing except that it is a chain of something called 'lookups'

> more ad-hoc code = more bugs, whereas functional expression = less code, clearer behavior. 

I'm not sure what you're referring to.  I like encapsulating behavior of functions in a class.

> I don't feel strongly about 3. although it seems a bit overkill at this stage.

agreed.  can we come to some agreement plz?  I'm trying.  Can you sell me on it with more details than saying it's more functional?